src/wallet/wallet2.cpp

```diff
@@ -5025,6 +5025,230 @@ bool wallet2::get_tx_key(const crypto::hash &txid, crypto::secret_key &tx_key) c
+bool wallet2::check_spend_proof(const crypto::hash& txid, const std::string& signatures_str)
+{
+  const size_t header_len = strlen("SpendProofV1");
+  THROW_WALLET_EXCEPTION_IF(signatures_str.size() < header_len || signatures_str.substr(0, header_len) != "SpendProofV1",
+    error::wallet_internal_error, "Signature header check error");
+
+  std::string decoded;
+  THROW_WALLET_EXCEPTION_IF(!tools::base58::decode(signatures_str.substr(header_len), decoded),
+    error::wallet_internal_error, "Signature decoding error");
+
+  // fetch tx from daemon
+  COMMAND_RPC_GET_TRANSACTIONS::request req = AUTO_VAL_INIT(req);
+  req.txs_hashes.push_back(epee::string_tools::pod_to_hex(txid));
+  req.decode_as_json = false;
+  COMMAND_RPC_GET_TRANSACTIONS::response res = AUTO_VAL_INIT(res);
+  m_daemon_rpc_mutex.lock();
+  bool r = epee::net_utils::invoke_http_json("/gettransactions", req, res, m_http_client, rpc_timeout);
+  m_daemon_rpc_mutex.unlock();
+  THROW_WALLET_EXCEPTION_IF(!r, error::no_connection_to_daemon, "gettransactions");
+  THROW_WALLET_EXCEPTION_IF(res.status == CORE_RPC_STATUS_BUSY, error::daemon_busy, "gettransactions");
+  THROW_WALLET_EXCEPTION_IF(res.status != CORE_RPC_STATUS_OK, error::wallet_internal_error, "gettransactions");
+  THROW_WALLET_EXCEPTION_IF(res.txs.size() != 1, error::wallet_internal_error,
+    "daemon returned wrong response for gettransactions, wrong txs count = " +
+    std::to_string(res.txs.size()) + ", expected 1");
+  cryptonote::blobdata bd;
+  THROW_WALLET_EXCEPTION_IF(!epee::string_tools::parse_hexstr_to_binbuff(res.txs[0].as_hex, bd), error::wallet_internal_error, "failed to parse tx from hexstr");
+  cryptonote::transaction tx;
+  crypto::hash tx_hash, tx_prefix_hash;
+  THROW_WALLET_EXCEPTION_IF(!cryptonote::parse_and_validate_tx_from_blob(bd, tx, tx_hash, tx_prefix_hash), error::wallet_internal_error, "failed to parse tx from blob");
+  THROW_WALLET_EXCEPTION_IF(tx_hash != txid, error::wallet_internal_error, "txid mismatch");
+
+  size_t count = 0;
+  for(size_t i = 0; i < tx.vin.size(); ++i)
+  {
+    if(tx.vin[i].type() != typeid(cryptonote::txin_to_key))
+      continue;
+    const txin_to_key& in_key = boost::get<txin_to_key>(tx.vin[i]);
+    count += in_key.key_offsets.size();
+  }
+  THROW_WALLET_EXCEPTION_IF(count * sizeof(crypto::signature) != decoded.size(),
+    error::wallet_internal_error, "incorrect signature size");
+
+  std::vector<std::vector<crypto::signature>> signatures;
+  count = 0;
+  for(size_t i = 0; i < tx.vin.size(); ++i)
+  {
+    if(tx.vin[i].type() != typeid(cryptonote::txin_to_key))
+      continue;
+    const txin_to_key& in_key = boost::get<txin_to_key>(tx.vin[i]);
+    signatures.push_back(std::vector<crypto::signature>{});
+    for (size_t j = 0; j < in_key.key_offsets.size(); ++j)
+    {
+      signatures.back().push_back(crypto::signature{});
+      memcpy(&signatures.back().back(), &decoded[count * sizeof(crypto::signature)], sizeof(crypto::signature));
```

Comment:
The default copy constructor is a bitwise `memove` so `signatures.back().push_back(decoded[count]);` should do what you want. Also these loops might be more readable with a C++11 range-based syntax; currently there are 3 different counters in here.

Reply:
Oops, just realized this was a raw byte array, nevermind.